### ISSUE TITLE
PlaceholderParser.js: caching makes problems

### ISSUE THREAD

**Issue by mnieper** [google/traceur-compiler#1712]
This is the beginning of the parse method in `PlaceholderParser.js`:

```javascript
function parse(sourceLiterals, values, doParse) {
  var tree = cache.get(sourceLiterals);
  if (!tree) {
    var source = insertPlaceholderIdentifiers(sourceLiterals);
    var errorReporter = new CollectingErrorReporter();
    var parser = getParser(source, errorReporter);
    tree = doParse(parser);
    if (errorReporter.hadError() || !tree || !parser.isAtEnd()) {
      throw new Error(
          `Internal error trying to parse:\n\n${source}\n\n${
              errorReporter.errorsAsString()}`);
    }
    cache.set(sourceLiterals, tree);
  }
```

In my current work on async generators, I am using this function (via `parseStatement` and friends). When I parse the same template literal twice (when I want to transpile more than one function), I run into a BlockBindingTransformer scope mismatch error. If I change `parse` so that the tree is not looked up in the cache but created anew on every invocation, the error goes away.

**Comment by mnieper** [comment]
P.S.: Just in case it helps, the scope mismatch is reported at the end of `BlockBindingTransformer.transformCatch`.

**Comment by arv** [comment]
The caching is important for speed. Maybe we could disable caching for when we need unique tree instances?

**Comment by domenic** [comment]
Here is my workaround for bypassing the caching: https://github.com/domenic/webidl-class-generator/blob/master/lib/traceur-parsing.js

I clone the tree after using the Traceur parser.

**Comment by mnieper** [comment]
Thanks for the workaround. The long term goal should be to ensure that transformers that need unique trees (as the BlockBindingTransformer) clone them automatically.
